14547: avoid evaluation of (Dimension|Quantity - 1)

smichr opened the pull request:
Issue #14547 has, at its root, a dimensional quantity being subtracted from 1 to see how it compares to 1. This raises an error for dimensional mismatch of addends. 

Addition of numbers (and expressions containing free symbols) to quantities or dimensions is now ignored but can be tested with the added `check_dimensions` function.

closes #14886 by replacement
fixes #14547
fixes #14932 

Better unification of units with the `quantity_simplify` function is now given, too.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. If there is no release notes entry for this PR,
write "NO ENTRY". The bot will check your release notes automatically to see
if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.units
  * checking for consistency of terms is no longer done at instantiation
  * `check_dimensions` was added to units/util.py to check expressions for dimensional consistency
  * `__radd__` and `__rmul__` are defined for Dimensions
  * `quantity_simplify` now reduces all Quantities in an Expr with the same Dimension to the same internally canonical Quantity
<!-- END RELEASE NOTES -->

TODO
- [x] see if `__rsub__` and `__rdiv__` need to be defined for Dimension
`__rsub__` is not needed because negation turns `-Dim` into `Dim`
- [x] keep sign of other when subtracting?
This is being left unchanged

smichr commented on sympy/physics/units/quantities.py:
```diff
@@ -21,9 +21,7 @@ class Quantity(AtomicExpr):
     """
 
     is_commutative = True
-    is_real = True
```
these are redundant (as added tests show)

asmeurer replied:
The only assumption I see that relates to this is `_eval_is_positive`, which could return False if the scale factor isn't positive (though I don't know if complex scale factors are possible).

asmeurer commented on sympy/functions/elementary/exponential.py:
```diff
@@ -693,22 +693,30 @@ def as_real_imag(self, deep=True, **hints):
+    @property
+    def _arg_is_1(self):
+        from sympy.physics.units import Quantity, Dimension
+        if self.has(Quantity, Dimension):
```
This sort of thing is disconcerting. Will everything in SymPy have to have guards against Quantity? 

I guess if we decide that `log` is really the only function that supports quantities, then it isn't a big deal.

smichr replied:
> This sort of thing is disconcerting

I know. I think this is a special situation for `log` because it is 0 at arg=1. So the assumptions test to see if it is 1 or not by computing `arg - 1`. An alternative would be to use a try/except to handle that error. Maybe that would be preferred to this current proposal?

smichr replied:
like this:
```
try:
 arg - 1
except ValueError as e:
 if 'incompatible' not in e.args[0]:
  raise e
```

Or should the error match be exact, `if 'summation of quantities of incompatible dimensions' != e.args[0]`?

asmeurer replied:
It's still a problem that the core has to know about Quantity. Even if we abstract away Quantity, it complicates the code everywhere if we can't even assume that expressions can be added to scalars.

schymans replied:
I think we may run into problems with other functions, too, such as trigonometric functions etc.
Maybe a totally crazy idea, but to avoid the burden of checking for quantities and units all the time, why don't we switch this off as default and just add a function `check_dimensions` to allow the user to check the dimensions of any expression at any point in time? This way, users could do whatever algebra they like with quantities, but in the end they will not get a meaningful result if the dimensions are not consistent. It would then be up to the user to algebraically resolve any units in functions to something meaningful before they perform a calculation: `log(1*meter)` would be unevaluated unless the user e.g. takes the difference: `log(1*meter) - log(1*centimeter) = log(1/0.01) = 4.6`

schymans replied:
In the interim, could we not add error handling to all of the `_eval_is_...` functions? They should only return `True` if the condition is actually true, whereas if it is not true or raises an error, as in the case of Quantities, the result should simply be `False`.

schymans reviewed (commented):
I think the idea is good, but instead of checking for Dimensions, I would just do something along the lines of:
```
try: 
     if (self.args[0] - 1).is_zero:
          return True
except:
     return False
return False
```

schymans commented:
I understand that bare `except` blocks are bad practice from the programming perspective, but for a condition "only if", I think that such a block is reasonable. If the condition cannot be interpreted for whatever reason, it certainly does not satisfy the "only if" condition. But having said this, I don't quite understand the necessity of the `(self.args[0] - 1)` constructs. Stupid question: Doesn't `(self.args[0] - 1).is_zero` return the same as `self.args[0] == 1`?

smichr committed:
14923: func(quantity) is not quantity

smichr commented:
> Stupid question: Doesn't (self.args[0] - 1).is_zero return the same as self.args[0] == 1?

If `self.args[0]` is a literal "1", yes. But the question about being zero or not can be answered for more than just 1.

```
>>> var('n',negative=1)
n
>>> (n-1).is_zero
False
>>> var('p',positive=1)
p
>>> (p-1).is_zero
>>> (1/(p+1)-1).is_zero
False
```

Actually, if this were testing for a non-literal one I'm not sure it would return anything by None so I am guessing it is there for the cases which are known *not* to be 1.

```
>>> one
1/(1 + pi) + pi/(1 + pi)
>>> (one - 1).is_zero
```

smichr committed:
14547: only raise errors with dimensional args

smichr committed:
quantity_simplify gives globally consistent units

smichr commented:
ping @schymans , @asmeurer

smichr committed:
make expression assertions more definate

smichr committed:
define rsub and keep args with free symbols

schymans reviewed (commented):
Does this mean that something like `1 + length` will not raise an exception any more? I still believe it should. Sorry, I forgot to refer to the respective line of code, see below.

schymans commented on sympy/physics/units/dimensions.py:
```diff
@@ -150,20 +151,17 @@ def _register_as_base_dim(self):
+             other = Dimension(Quantity.get_dimensional_expr(other))
+        if self == other:
+            return self
+        elif isinstance(other, Dimension):
```
Does this mean that the exception is only raised if both, self and other are dimensions, i.e. not for e.g. `1 + length`?

smichr replied:
yes, for the same reason as for allowing `1 + foot` (see answer below for motivation)

schymans commented on sympy/physics/units/quantities.py:
```diff
@@ -252,15 +252,18 @@ def free_symbols(self):
+    # expressions like `meter + second` to be created. For sake of
+    # core operations involving assumptions, the case of `unit + number`
+    # will be ignored.
 
```
I still believe that `unit + number` should raise an exception.

Upabjojr replied:
I think so as well.

schymans commented on sympy/physics/units/tests/test_dimensions.py:
```diff
@@ -65,17 +66,23 @@ def test_Dimension_properties():
 
-    raises(TypeError, lambda: length + 1)
-    raises(TypeError, lambda: length - 1)
     raises(ValueError, lambda: length + time)
```
I believe that `length + 1` should raise an exception in any dimension system, whereas `length + time` should not always, e.g. if we use natural units: https://github.com/sympy/sympy/issues/14374

schymans commented on sympy/physics/units/tests/test_quantities.py:
```diff
@@ -318,7 +318,6 @@ def test_Quantity_derivative():
 
 def test_sum_of_incompatible_quantities():
-    raises(ValueError, lambda: meter + 1)
     raises(ValueError, lambda: meter + second)
```
See my comment above.

schymans reviewed (commented):
I think it is great, but I do not understand the necessity to allow e.g. `length + 1` to pass. Could you elaborate?

smichr commented:
> Could you elaborate?

The assumption system in the core assumes that you can subtract 1 from an expression. If we raise an error when doing so then we have to make the core aware of this possibility and it slows every object down because of having to do this check.

Also, the beauty of CAS is that if someone makes a dumb mistake (like adding 1 to a value with units) that mistake is not lost and at the end of a computation one might have "3*meters + 1"...and they can ask *themselves* what that means. SymPy will simply record how they "shot themselves in the foot." 

So if SymPy becomes a nanny for these operations and prohibits them, then such objects will have only a limited use. If you want robust interactions with the object having units/dimensions then they have to be able to deal with what the core will do to them, for example, in doing solution checking and passing through the assumption system.

As far as the `foot + meter` issue...when we have a natural unit system defined (and it is my understanding that we do not yet have this) then we can worry about this issue. Is there any reason we shouldn't allow this at this time?

schymans commented:
I understand, thanks. However, if `1 + length` does not raise an exception, then `time + length` should also not raise an exception, otherwise we are inconsistent and confusing. Along the same line of argument that you propose above, we should allow the users to add all possible dimensions together and figure out at the end if the result makes sense. There is nothing worse than a false sense of security, i.e. protection from one kind of meaningless interaction but not another. Would we break anything if we removed all the exceptions from `__add__` and moved them to a function `check_dimensions`?

schymans commented:
Actually, thinking about it again, I think indeed that returning `length + time` is more informative than:
```python
ValueError: summation of quantities of incompatible dimensions
```
In `check_dimensions`, we could just run `simplify` on an expression and if the result is of type `Add`, we could raise the above ValueError and provide the dimensional expression. Would this help?

smichr commented:
> Would we break anything if we removed all the exceptions from __add__ and moved them to a function check_dimensions?

This would be my preference. I'll see what I can do.

asmeurer commented:
I agree with @smichr here. If the units want to be Expr objects, they need to be able to be added to other Expr objects. Stuff like the `self.args[0] - 1` that log does could appear anywhere in the code. So you either need to

1. Make it so that quantities and other Expr objects can be combined without issues. There can be a function that does a type check on quantities. Regarding combining quantities with other quantities I'm not sure but I think having it check at construction time should be fine. 

2. Make quantities non-Expr, meaning that they can't be combined with arbitrary expressions. Special things like expr*quantity or log(quantity) would need to be supported explicitly (similar to how matrices work).

schymans commented:
@asmeurer, if we opt for your second option, would this open the way to a better separation of numbers and units, e.g. `3 * kilogram * meter /(2 * second ** 2) == 3/2*kg*m/s**2` instead of `3*kilogram*meter/(2*second**2)`? 
I would be in favour of this.

smichr committed:
clear mul constants in check_dimensions

smichr committed:
remove Dimension postprocessing for Add

smichr commented:
> I would be in favour of this.

OK, maybe that could be an separate change. Can we commit this current PR as is and let someone else take up making the Dimension/Quantity non-Expr? This PR at least corrects the current troubles with inequalities.

schymans commented on sympy/physics/units/util.py:
```diff
@@ -132,7 +132,7 @@ def get_total_scale_factor(expr):
 def quantity_simplify(expr):
     """Return an equivalent expression in which prefixes are replaced
-    with numerical values and products of related quantities are
+    with numerical values and all units of a given dimension are the
```
I think you need to remove "the" at the end of the line.

schymans commented on sympy/physics/units/tests/test_dimensions.py:
```diff
@@ -65,17 +66,33 @@ def test_Dimension_properties():
 
-    raises(TypeError, lambda: length + 1)
-    raises(TypeError, lambda: length - 1)
     raises(ValueError, lambda: length + time)
```
I thought that dimensions are not checked for consistency any more, unless specified, so `length + time` should not raise an error.

smichr replied:
hmm...good catch. Not intentionally left behind. I will look into it.

schymans reviewed (changes requested):
I think this is great, but I am confused about the remaining `raises(ValueError, lambda: length + time)`. I thought that removal of the `_Quantity_constructor_postprocessor_Add` would remove the exception.

smichr committed:
remove checking of Dimension compatibility
